### PR TITLE
Throw error on aborted rebase

### PR DESCRIPTION
**Pull request by simonbyrne**
`LibGit2.rebase!` will no longer silently abort if an error occurs.

This changes the behaviour of one of the tests: for some reason LibGit2 is unable to automatically rebase this branch (previously it would silently abort).

follows on from #19624

### CODE REVIEW

**Comment by tkelman** [comment]
what's the error? does it need a change in merge options or something?

**Comment by simonbyrne** [comment]
It gives `EUNMERGED` (as it tries to make a commit with unmerged changes).

I'm not sure if it can be fixed by different merge options.

**Comment by tkelman** [comment]
any idea why command line git works but libgit2 doesn't then? is this a bug in libgit2?

**Comment by simonbyrne** [comment]
Okay, it seemed to be due to a problem with how we were setting up remote tracking branches in the test case. Unfortunately fixing this exposed another problem with how we handle merge when the tracking branch has no commits. Anyway, this should fix them both.

File: base/libgit2/libgit2.jl [hunk 1]
```diff
@@ -386,46 +386,68 @@ function merge!(repo::GitRepo;
...
+                # this isn't really a merge, but really moving HEAD
+                # https://github.com/libgit2/libgit2/issues/2135#issuecomment-35997764
+                # try to figure out remote tracking of unborn head
+
+                m = with(GitReference(repo, Consts.HEAD_FILE)) do head_sym_ref
+                    match(r"refs/heads/(.*)", fullname(head_sym_ref))
                 end
-            else # try to get tracking remote branch for the head
-                if !isattached(repo)
+                if m == nothing
```
Comment (tkelman) [inline base/libgit2/libgit2.jl]: `===`

Comment (simonbyrne) [inline base/libgit2/libgit2.jl]: Any reason why?

Comment (tkelman) [inline base/libgit2/libgit2.jl]: nothing is a singleton, so it just cuts out one level of inlining

File: base/libgit2/libgit2.jl [hunk 2]
```diff
@@ -424,9 +424,11 @@ function merge!(repo::GitRepo;
                 remotename = with(GitConfig, repo) do cfg
                     LibGit2.get(String, cfg, "branch.$branchname.remote")
                 end
-                with(GitReference(repo, "refs/remotes/$remotename/$branchname")) do ref
-                    obj = LibGit2.Oid(ref)
-                    LibGit2.create_branch(repo, branchname, get(GitCommit, repo, obj))
+                obj = with(GitReference(repo, "refs/remotes/$remotename/$branchname")) do ref
+                    LibGit2.Oid(ref)
+                end
+                with(get(GitCommit, repo, obj)) do cmt
```
Comment (tkelman) [inline base/libgit2/libgit2.jl]: have you been able to come up with any scheme or logic for when and what needs management like this?

**Comment by StefanKarpinski** [comment]
Bump?

File: base/libgit2/libgit2.jl [hunk 3]
```diff
@@ -386,51 +386,91 @@ function merge!(repo::GitRepo;
...
+        merge!(repo, upst_anns, fastforward,
+               merge_opts=merge_opts,
+               checkout_opts=checkout_opts)
+    finally
+        map(finalize, upst_anns)
     end
 end
 
-""" git rebase --merge [--onto <newbase>] [<upstream>] """
-function rebase!(repo::GitRepo, upstream::AbstractString="", newbase::AbstractString="")
```
Comment (tkelman) [inline base/libgit2/libgit2.jl]: was newbase not hooked up at all?

Comment (simonbyrne) [inline base/libgit2/libgit2.jl]: No

Comment (tkelman) [inline base/libgit2/libgit2.jl]: guess we should add that to the list in #19839 since it would be a useful feature to have at some point

Comment (simonbyrne) [inline base/libgit2/libgit2.jl]: good point.

**Comment by tkelman** [comment]
It should be safe to ignore the osx travis failure here. It and linux 64 both passed last time this ran a few weeks ago, I restarted those two jobs just in case anything on master would have caused issues when this is merged. The osx travis issue is just that changes to `.travis.yml` don't always take effect properly on a restart (here for the new LLVM version).

**Pull request merged**